File: implementations/quicksort.py
import time
def quicksort(arr):
    if len(arr) <= 1:
        return arr
    pivot = arr[len(arr) // 2]
    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]
    return quicksort(left) + middle + quicksort(right)

start = time.time()
quicksort([5,2,7,9,4,2,6,1,6,9])
end = time.time()
print(f"Time taken: {end - start}")

File: implementations/quicksort.go
package main

import (
	"fmt"
	"time"
	// "math/rand"
)

func quickSort(arr []int) []int {
	middle := arr[len(arr)-1]
	var smaller []int
	var larger []int
	numMiddles := 0
	for _, e := range arr {
		if e < middle {
			smaller = append(smaller, e)
		} else if e > middle {
			larger = append(larger, e)
		} else if e == middle {
			numMiddles += 1
		}
	}
	if len(smaller) > 1 {
		smaller = quickSort(smaller)
	}
	if len(larger) > 1 {
		larger = quickSort(larger)
	}
	arr = append(smaller, middle)

	for i := 0; i < numMiddles-1; i++ {
		arr = append(arr, middle)
	}
	for i := 0; i < len(larger); i++ {
		arr = append(arr, larger[i])
	}
	return arr
}

func main() {
	randarr := []int{9,3,6,2,8,7,3,4}
	start := time.Now()
	quickSort(randarr)
	fmt.Println(time.Since(start))
}
